### PR TITLE
[refactor] Compile the Ndarray argument to a struct

### PR DESCRIPTION
**Pull request by lin-hitonami**
**In this PR we represent an ndarray with type `tp` argument with element type `tp` as` struct{primal: *tp}`. We will add fields to the struct in following prs. It will become `struct{shape: array(n), primal: *tp, grad: *tp}` in the end.**

**We may add an ndarray type in CHI-IR and lower it to struct type in codegen in the future, but we temporarily represent the ndarray type with struct type in CHI-IR in this PR.**

### Summary
🔄🐛🆕

This pull request refactors the ndarray support in Taichi, making ndarrays first-class citizens in the IR and kernel representations. It introduces new helper functions, such as `create_ndarray_arg_load` and `insert_ndarray_param`, to simplify the creation and insertion of ndarrays as kernel arguments and parameters. It also updates the code generation and runtime logic for various backends, such as LLVM, SPIRV, and CUDA, to handle ndarrays correctly. It also modifies the tests and the Python interface to reflect the new ndarray support.

> _`ndarray` is the new way to pass the data_
> _Structs and pointers are the keys to the gate_
> _We refactor the code, we generate the IR_
> _We unleash the power of the kernel fire_

### Walkthrough
*  Refactor the kernel arguments to support ndarrays as first-class citizens in Taichi ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-575efc738df7b1202370c2531ec82232dc7f287b2bec4999af03ef40da4f5deeL90-R90), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-a6e92dd2dd707d705dc44ef91463ddc0423575188e6e8e0555de9e439db88c35L599-R607), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-bdb4f85a29d6478a4482d81ca072237534fb641b52f3c529aca93e872ade6fecL495-R504), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-1894085b261e833e3e66924fc5b1cf63b9dd8b8aa0b3e78ec64366396131470dR148-R149), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-8505840874bba49dcec6e2a71c82e0fea5cfc9ab1f0194836f022fc5a8fa2155R32-R48), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-8505840874bba49dcec6e2a71c82e0fea5cfc9ab1f0194836f022fc5a8fa2155L76-R93), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-e914f7cb94f429034d998974d95d57f6e09de2dadaed4c621e122ec403c9bb22R116-R119), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-af631a0c71978fe591e17005f01f7c06bc30ae36c65df306bbb3b08ade770167R679), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-97b0d9ab204b703802b3b5d04d036d30f66b34b726128216faf0d8a2a8564528L838-R841), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-dd572dab7be4dbb5edc1043d6d6339b931ef35198b8657761ebf45a83e76ac2bL465-R467), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-524d02dcd4593680562069f2c0566585c3976164593989ed4aff0657b8c06ee8L114-R114), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-524d02dcd4593680562069f2c0566585c3976164593989ed4aff0657b8c06ee8L175-R175), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-8077cae3c0b83bc7f6faad3d1e62a35ec53ebe876f0686706b117c5c4f5a4321L24-R24), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-8077cae3c0b83bc7f6faad3d1e62a35ec53ebe876f0686706b117c5c4f5a4321L44-R44))
  * Rename the function `insert_arr_param` to `insert_ndarray_param` in `kernel_arguments.py` and `callable.h` ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-575efc738df7b1202370c2531ec82232dc7f287b2bec4999af03ef40da4f5deeL90-R90), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-e914f7cb94f429034d998974d95d57f6e09de2dadaed4c621e122ec403c9bb22R116-R119))
  * Add a new function `create_ndarray_arg_load` to the `IRBuilder` class in `ir_builder.cpp` and `ir_builder.h` that creates an `ArgLoadStmt` with a struct type that contains a pointer type as a member ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-bdb4f85a29d6478a4482d81ca072237534fb641b52f3c529aca93e872ade6fecL495-R504), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-1894085b261e833e3e66924fc5b1cf63b9dd8b8aa0b3e78ec64366396131470dR148-R149))
  * Use the new function `create_ndarray_arg_load` to create `ArgLoadStmt` for ndarrays in `frontend_ir.cpp`, `ir_builder_test.cpp`, `ndarray_kernel.cpp`, and `half2_vectorization_test.cpp` ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-a6e92dd2dd707d705dc44ef91463ddc0423575188e6e8e0555de9e439db88c35L599-R607), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-524d02dcd4593680562069f2c0566585c3976164593989ed4aff0657b8c06ee8L99-R100), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-524d02dcd4593680562069f2c0566585c3976164593989ed4aff0657b8c06ee8L167-R168), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-8077cae3c0b83bc7f6faad3d1e62a35ec53ebe876f0686706b117c5c4f5a4321L8-R9), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-8077cae3c0b83bc7f6faad3d1e62a35ec53ebe876f0686706b117c5c4f5a4321L34-R35), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-5136d70f7a32456bee3938daca1066aa3d380aecd7d18257fc893b83dfd72a79L61-R69))
  * Add a new function `insert_ndarray_param` to the `Callable` class in `callable.cpp` and `callable.h` that inserts a new parameter of type ndarray to the parameter list ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-8505840874bba49dcec6e2a71c82e0fea5cfc9ab1f0194836f022fc5a8fa2155R32-R48), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-e914f7cb94f429034d998974d95d57f6e09de2dadaed4c621e122ec403c9bb22R116-R119))
  * Use the new function `insert_ndarray_param` to insert array parameters for ndarrays in `ir_builder_test.cpp` and `ndarray_kernel.cpp` ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-524d02dcd4593680562069f2c0566585c3976164593989ed4aff0657b8c06ee8L114-R114), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-524d02dcd4593680562069f2c0566585c3976164593989ed4aff0657b8c06ee8L175-R175), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-8077cae3c0b83bc7f6faad3d1e62a35ec53ebe876f0686706b117c5c4f5a4321L24-R24), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-8077cae3c0b83bc7f6faad3d1e62a35ec53ebe876f0686706b117c5c4f5a4321L44-R44))
  * Modify the condition for creating a pointer type for the parameter type in the `finalize_params` function of the `Callable` class in `callable.cpp` to check if the parameter type is a struct type that contains a pointer type as a member ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-8505840874bba49dcec6e2a71c82e0fea5cfc9ab1f0194836f022fc5a8fa2155L76-R93))
  * Add a new binding for the `insert_ndarray_param` function of the `Kernel` class to the Python interface in `export_lang.cpp` ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-af631a0c71978fe591e17005f01f7c06bc30ae36c65df306bbb3b08ade770167R679))
  * Modify the assignment of the `ret_type` of the `base_ptr` statement in the `visit` function of the `MergeExternalAndMatrixPtr` class in `scalarize.cpp` to create a struct type with a pointer type as a member ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-97b0d9ab204b703802b3b5d04d036d30f66b34b726128216faf0d8a2a8564528L838-R841))
  * Modify the assignment of the `ret_type` of the `stmt` statement in the `visit` function of the `TypeCheck` class in `type_check.cpp` to extract the pointer type from the struct type of the `base_ptr` statement ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-dd572dab7be4dbb5edc1043d6d6339b931ef35198b8657761ebf45a83e76ac2bL465-R467))
* Implement the LLVM backend for ndarrays ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-3c663c78745adcd3f6a7ac81fe99e628decc3040f292ea1e20ecd4b85a7f4313R1846-R1862), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-3c663c78745adcd3f6a7ac81fe99e628decc3040f292ea1e20ecd4b85a7f4313L1905-R1923), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-3c663c78745adcd3f6a7ac81fe99e628decc3040f292ea1e20ecd4b85a7f4313L1931-R1944), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-3c663c78745adcd3f6a7ac81fe99e628decc3040f292ea1e20ecd4b85a7f4313L1938-R1951), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-e2ee5e63fdd0c54971da27aa847272d52e9f74cec28413cf6c0a1c6f66cc14aeL21-R21), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-e2ee5e63fdd0c54971da27aa847272d52e9f74cec28413cf6c0a1c6f66cc14aeL30-R30), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-d8cf951dea16cc738fc36faee3a158f37a605abccd312c4a6f18d73fd2504332L65-R65), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-d8cf951dea16cc738fc36faee3a158f37a605abccd312c4a6f18d73fd2504332L80-R80), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-5de6a93ab8620f5ececa55c3e762efa567d428c85058593ab49bb641f0c4c2fcL84-R84))
  * Add a new case to handle the `PtrOffsetStmt` when the base pointer is an ndarray argument in `codegen_llvm.cpp` ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-3c663c78745adcd3f6a7ac81fe99e628decc3040f292ea1e20ecd4b85a7f4313R1846-R1862))
  * Remove the redundant code that casts the base pointer to a primitive pointer type in `codegen_llvm.cpp` ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-3c663c78745adcd3f6a7ac81fe99e628decc3040f292ea1e20ecd4b85a7f4313L1905-R1923))
  * Replace the use of `stmt->base_ptr` with `ptr_val`, which is the loaded value of the data pointer member of the struct type, in `codegen_llvm.cpp` ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-3c663c78745adcd3f6a7ac81fe99e628decc3040f292ea1e20ecd4b85a7f4313L1931-R1944), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-3c663c78745adcd3f6a7ac81fe99e628decc3040f292ea1e20ecd4b85a7f4313L1938-R1951))
  * Replace the call to the `set_arg` function with a call to the `set_struct_arg` function, and pass a vector with the argument index and zero as the argument indices, in the `launch_llvm_kernel` function of the CPU kernel launcher in `kernel_launcher.cpp`, the CUDA kernel launcher in `kernel_launcher.cpp`, and the `host_to_device` function of the `HostDeviceContextBlitter` class in `runtime.cpp` ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-e2ee5e63fdd0c54971da27aa847272d52e9f74cec28413cf6c0a1c6f66cc14aeL21-R21), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-e2ee5e63fdd0c54971da27aa847272d52e9f74cec28413cf6c0a1c6f66cc14aeL30-R30), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-d8cf951dea16cc738fc36faee3a158f37a605abccd312c4a6f18d73fd2504332L65-R65), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-d8cf951dea16cc738fc36faee3a158f37a605abccd312c4a6f18d73fd2504332L80-R80), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-5de6a93ab8620f5ececa55c3e762efa567d428c85058593ab49bb641f0c4c2fcL84-R84))
* Implement the SPIRV backend for ndarrays ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-1620f2a387fc8acc55e2b2cfced07bb9cba59702609aae6e9489e703cbab5000L570-R573), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-1620f2a387fc8acc55e2b2cfced07bb9cba59702609aae6e9489e703cbab5000L695-R698), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-1620f2a387fc8acc55e2b2cfced07bb9cba59702609aae6e9489e703cbab5000L706-R710), [link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-5274ca34a1156d70b415ce034aeea8430182b5b15b95b3cef9efa79661028e15R353-R355))
  * Add a new condition to check if the argument type is a struct type that contains a pointer type as a member in `spirv_codegen.cpp` ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-1620f2a387fc8acc55e2b2cfced07bb9cba59702609aae6e9489e703cbab5000L570-R573))
  * Replace the use of `argload->ret_type` with `stmt->ret_type` in `spirv_codegen.cpp` ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-1620f2a387fc8acc55e2b2cfced07bb9cba59702609aae6e9489e703cbab5000L695-R698))
  * Add a new index to the GEP operation that accesses the data pointer member of the struct type of the argument in `spirv_codegen.cpp` ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-1620f2a387fc8acc55e2b2cfced07bb9cba59702609aae6e9489e703cbab5000L706-R710))
  * Add a new assertion to check if the data type is a primitive type in `spirv_ir_builder.cpp` ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-5274ca34a1156d70b415ce034aeea8430182b5b15b95b3cef9efa79661028e15R353-R355))
* Delete the file `test_ad_ndarray_cpu.py` ([link](https://github.com/taichi-dev/taichi/pull/7809/files?diff=unified&w=0#diff-98252c69b985843ca3b701283b70dc21d11a0728963cf60bc10d56f4f7834c1d))

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #7849
* __->__ #7809
* #7788

### CODE REVIEW

File: taichi/codegen/llvm/codegen_llvm.cpp [hunk 1]
```diff
@@ -1843,6 +1843,23 @@ void TaskCodeGenLLVM::visit(MatrixPtrStmt *stmt) {
...
+                               ->get_pointee_type();
+  auto arg_type = operand_dtype;
+  if (operand_dtype->is<TensorType>()) {
+    arg_type = operand_dtype->as<TensorType>()->get_element_type();
+  }
+  auto ptr_type = TypeFactory::get_instance().get_pointer_type(arg_type);
+  auto *struct_type = tlctx->get_data_type(
+      TypeFactory::get_instance().get_struct_type({{ptr_type}}));
+  std::vector<llvm::Value *> index(2, tlctx->get_constant(0));
+  auto *gep = builder->CreateGEP(struct_type, llvm_val[stmt->base_ptr], index);
```
Comment (bobcao3) [inline taichi/codegen/llvm/codegen_llvm.cpp:1860]: Is `llvm_val` a map? If so use `.at()`

File: taichi/codegen/spirv/spirv_codegen.cpp [hunk 2]
```diff
@@ -567,7 +567,10 @@ class TaskCodegen : public IRVisitor {
   void visit(ArgLoadStmt *stmt) override {
     const auto arg_id = stmt->arg_id;
     const auto arg_type = ctx_attribs_->args_type()->get_element_type({arg_id});
-    if (arg_type->is<PointerType>()) {
+    if (arg_type->is<PointerType>() ||
+        (arg_type->is<lang::StructType>() && arg_type->as<lang::StructType>()
```
Comment (bobcao3) [inline taichi/codegen/spirv/spirv_codegen.cpp:571]: I think we shouldn't add NDarray type (complicates IR), but rather rely on annotation information or the context buffer argument type.

**Review by bobcao3 — APPROVED** [review]
LGTM other than the nits

File: taichi/codegen/spirv/spirv_codegen.cpp (hunk 2 above)
Comment (lin-hitonami) [reply taichi/codegen/spirv/spirv_codegen.cpp:571]: You mean we should use `KernelContextAttributes::arg_attribs_vec_` or `Callable::parameter_list`? (btw we should unify them)

**Pull request closed without merging**